File: julia/2015/day07a/day07a.jl
#!/usr/bin/env julia

struct Assign
    src :: String
end

struct Not
    src :: String
end

struct And
    src1 :: String
    src2 :: String
end

struct Or
    src1 :: String
    src2 :: String
end

struct LeftShift
    src :: String
    amt :: Int32
end

struct RightShift
    src :: String
    amt :: Int32
end

const Operation = Union{Assign, Not, And, Or, LeftShift, RightShift}

function usage()
    progname = PROGRAM_FILE
    println(stderr, "usage: $progname <input file>")
    exit(1)
end

function process(content::String)::Int32
    operations = Dict{String, Operation}()
    cache = Dict{String, Int32}()
    src1 :: String = ""
    src2 :: String = ""
    amt :: Int32 = 0
    dest :: String = ""
    for line in split(content, '\n', keepempty=false)
        matches = match(r"^(\d+|\w+) -> (\w+)$", line)
        if matches != nothing
            src1 = matches.captures[1]
            dest = matches.captures[2]
            operations[dest] = Assign(src1)
        end
        matches = match(r"NOT (\d+|\w+) -> (\w+)", line)
        if matches != nothing
            src1 = matches.captures[1]
            dest = matches.captures[2]
            operations[dest] = Not(src1)
        end
        matches = match(r"(\d+|\w+) (AND|OR) (\d+|\w+) -> (\w+)", line)
        if matches != nothing
            src1 = matches.captures[1]
            op = matches.captures[2]
            src2 = matches.captures[3]
            dest = matches.captures[4]
            if op == "AND"
                operations[dest] = And(src1, src2)
            else
                operations[dest] = Or(src1, src2)
            end
        end
        matches = match(r"(\d+|\w+) (LSHIFT|RSHIFT) (\d+) -> (\w+)", line)
        if matches != nothing
            src1 = matches.captures[1]
            op = matches.captures[2]
            amt = parse(Int32, matches.captures[3])
            dest = matches.captures[4]
            if op == "LSHIFT"
                operations[dest] = LeftShift(src1, amt)
            else
                operations[dest] = RightShift(src1, amt)
            end
        end
    end
    return eval(operations, cache, "a")
end

function eval(ops, cache, expr)
    if tryparse(Int32, expr) !== nothing
        return parse(Int32, expr)
    elseif haskey(cache, expr)
        return cache[expr]
    else
        r = 0
        op = ops[expr]
        if op isa Assign
            r = eval(ops, cache, op.src)
        elseif op isa Not
            a = eval(ops, cache, op.src)
            r = ~a
        elseif op isa And
            a = eval(ops, cache, op.src1)
            b = eval(ops, cache, op.src2)
            r = a & b
        elseif op isa Or
            a = eval(ops, cache, op.src1)
            b = eval(ops, cache, op.src2)
            r = a | b
        elseif op isa LeftShift
            a = eval(ops, cache, op.src)
            r = a << op.amt
        elseif op isa RightShift
            a = eval(ops, cache, op.src)
            r = a >> op.amt
        end
        masked = r & 0xffff
        cache[expr] = masked
        return masked
    end
end

function main()
    args = ARGS
    if length(args) < 1
        usage()
    end
    filename = args[1]
    try
        content = read(filename, String)
        result = process(content)
        println("result = $result")
    catch e
        println(stderr, "error reading file: $e")
        exit(1)
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: nim/2015/day07a/day07a.nim
import options, os, re, strutils, tables

type
  Operator = enum
    Assign, Not, And, Or, LeftShift, RightShift

type
  Operation = object
    operator: Operator
    source1: string
    source2: Option[string] = none(string)
    amount: Option[int] = none(int)

proc eval(ops: TableRef[string, Operation], cache: TableRef[string, int], expr: string): int =
  if match(expr, re"\d+"):
    parseInt(expr)
  elif cache.hasKey(expr):
    cache[expr]
  else:
    let op = ops[expr]
    var r: int;
    case op.operator:
    of Assign:
      let a = eval(ops, cache, op.source1)
      r = a
    of Not:
      let a = eval(ops, cache, op.source1)
      r = not a
    of And:
      let a = eval(ops, cache, op.source1)
      let b = eval(ops, cache, op.source2.get())
      r = a and b
    of Or:
      let a = eval(ops, cache, op.source1)
      let b = eval(ops, cache, op.source2.get())
      r = a or b
    of LeftShift:
      let a = eval(ops, cache, op.source1)
      r = a shl op.amount.get()
    of RightShift:
      let a = eval(ops, cache, op.source1)
      r = a shr op.amount.get()
    let masked = r and 0xffff;
    cache[expr] = masked
    masked

proc processFile(filename: string): int =
  var operations = newTable[string, Operation]()
  var cache = newTable[string, int]()
  var caps: array[4, string];
  for line in lines(filename):
    if find(line, re"^(\d+|\w+) -> (\w+)$", caps) != -1:
      let src1 = caps[0]
      let dest = caps[1]
      operations[dest] = Operation(operator: Assign, source1: src1)
    elif find(line, re"NOT (\d+|\w+) -> (\w+)", caps) != -1:
      let src1 = caps[0]
      let dest = caps[1]
      operations[dest] = Operation(operator: Not, source1: src1)
    elif find(line, re"(\d+|\w+) (AND|OR) (\d+|\w+) -> (\w+)", caps) != -1:
      let src1 = caps[0]
      let op = caps[1]
      let src2 = caps[2]
      let dest = caps[3]
      operations[dest] = Operation(operator: (if op == "AND": And else: Or), source1: src1, source2: some(src2))
    elif find(line, re"(\d+|\w+) (LSHIFT|RSHIFT) (\d+) -> (\w+)", caps) != -1:
      let src1 = caps[0]
      let op = caps[1]
      let amt = parseInt(caps[2])
      let dest = caps[3]
      operations[dest] = Operation(operator: (if op == "LSHIFT": LeftShift else: RightShift), source1: src1, amount: some(amt))
  eval(operations, cache, "a")

when isMainModule:
  if paramCount() < 1:
    quit "usage: " & getAppFilename() & " <input file>", 1

  let filename = paramStr(1)
  try:
    echo "result = ", processFile(filename)
  except OSError as e:
    quit "error while processing file `" & filename & "': " & e.msg, 1
